server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java

```diff
@@ -76,37 +109,228 @@ public void onFailure(Exception e) {
             }));
     }
 
-    public void sendPublishRequest(DiscoveryNode destination, PublishRequest publishRequest,
-                                      ActionListener<PublishWithJoinResponse> responseActionListener) {
-        // TODO: serialize and compress state similar as in PublishClusterStateAction
-        TransportRequestOptions options = TransportRequestOptions.builder().withType(TransportRequestOptions.Type.STATE).build();
-        transportService.sendRequest(destination, PUBLISH_STATE_ACTION_NAME, publishRequest, options,
-            new TransportResponseHandler<PublishWithJoinResponse>() {
+    public PublishClusterStateStats stats() {
+        return new PublishClusterStateStats(
+            fullClusterStateReceivedCount.get(),
+            incompatibleClusterStateDiffReceivedCount.get(),
+            compatibleClusterStateDiffReceivedCount.get());
+    }
 
-                @Override
-                public PublishWithJoinResponse read(StreamInput in) throws IOException {
-                    return new PublishWithJoinResponse(in);
-                }
+    public interface PublicationContext {
 
-                @Override
-                public void handleResponse(PublishWithJoinResponse response) {
-                    responseActionListener.onResponse(response);
-                }
+        void sendPublishRequest(DiscoveryNode destination, PublishRequest publishRequest,
+                                ActionListener<PublishWithJoinResponse> responseActionListener);
 
-                @Override
-                public void handleException(TransportException exp) {
-                    responseActionListener.onFailure(exp);
+    }
+
+    public PublicationContext newPublicationContext(ClusterChangedEvent clusterChangedEvent) {
+        final DiscoveryNodes nodes = clusterChangedEvent.state().nodes();
+        final ClusterState newState = clusterChangedEvent.state();
+        final ClusterState previousState = clusterChangedEvent.previousState();
+        final boolean sendFullVersion = clusterChangedEvent.previousState().getBlocks().disableStatePersistence();
```

Comment:
Should we also obey `discovery.zen.publish_diff.enable` here?

Reply:
I would like to avoid exposing this as a setting and having this configurable in Zen2, which is why I deliberately opted not to implement this.